third_party/move/move-vm/runtime/src/data_cache.rs

```diff
@@ -70,7 +70,24 @@ impl<'r> TransactionDataCache<'r> {
     ///
     /// Gives all proper guarantees on lifetime of global data as well.
     pub(crate) fn into_effects(self, loader: &Loader) -> PartialVMResult<(ChangeSet, Vec<Event>)> {
-        let mut change_set = ChangeSet::new();
+        let resource_converter =
+            |value: Value, layout: MoveTypeLayout| -> PartialVMResult<Vec<u8>> {
+                value.simple_serialize(&layout).ok_or_else(|| {
+                    PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)
+                        .with_message(format!("Error when serializing resource {}.", value))
+                })
+            };
+        self.into_custom_effects(&resource_converter, loader)
+    }
+
+    /// Same like `into_effects`, but also allows clients to select the format of
+    /// produced effects for resources.
+    pub(crate) fn into_custom_effects<R>(
+        self,
+        resource_converter: &dyn Fn(Value, MoveTypeLayout) -> PartialVMResult<R>,
+        loader: &Loader,
+    ) -> PartialVMResult<(Changes<Vec<u8>, R>, Vec<Event>)> {
```

Comment:
so we still hard code M to Vec\<u8\> here?

Reply:
yes, unless we want to also have `CompiledModule`s or other type out?

Reply:
I think the new loader may need a different form for modules but no one knows for sure right now